sdk/core/core-rest-pipeline/src/util/tokenCycler.ts

```diff
@@ -52,50 +44,61 @@ export const DEFAULT_CYCLER_OPTIONS: TokenCyclerOptions = {
  *
- * @param getAccessToken - A function that produces a promise of an access token that may fail by returning null.
- * @param retryIntervalInMs - The time (in milliseconds) to wait between retry attempts.
- * @param refreshTimeout - The timestamp after which the refresh attempt will fail, throwing an exception.
- * @returns - A promise that, if it resolves, will resolve with an access token.
+ * @param asyncFn - The asynchronous function to call.
+ * @param timeout - Unix time when to stop ignoring errors or null values.
+ * @param errorMessage - Message to be used if the final value after the timeout is reached is still "null".
+ * @returns - A promise that can be resolved with T or null until the time is reached, after which it will throw if it can't resolve with a value of the type T.
  */
-async function beginRefresh(
-  getAccessToken: () => Promise<AccessToken | null>,
-  retryIntervalInMs: number,
-  refreshTimeout: number
-): Promise<AccessToken> {
-  // This wrapper handles exceptions gracefully as long as we haven't exceeded
-  // the timeout.
-  async function tryGetAccessToken(): Promise<AccessToken | null> {
-    if (Date.now() < refreshTimeout) {
-      try {
-        return await getAccessToken();
-      } catch {
-        return null;
-      }
-    } else {
-      const finalToken = await getAccessToken();
-
-      // Timeout is up, so throw if it's still null
-      if (finalToken === null) {
-        throw new Error("Failed to refresh access token.");
-      }
-
-      return finalToken;
+async function tryUntilTimeout<T>(
+  asyncFn: () => Promise<T | null>,
+  timeout: number,
+  errorMessage: string
+): Promise<T | null> {
+  if (Date.now() < timeout) {
+    try {
+      return asyncFn();
+    } catch {
+      return null;
+    }
+  } else {
+    const result = await asyncFn();
+
+    // Timeout is up, so throw if it's still null
+    if (result === null) {
+      throw new Error(errorMessage);
     }
+    return result;
   }
+}
 
-  let token: AccessToken | null = await tryGetAccessToken();
+/**
+ * Retries an asynchronous function as often as the provided interval milliseconds,
+ * until the provided Unix date timeout is reached.
+ * Once the timeout is reached, errors will bubble up, and null return values will throw with the received "errorMessage".
+ *
+ * @param asyncFn - The asynchronous function to call.
+ * @param retryIntervalInMs - The time (in milliseconds) to wait between retry attempts.
+ * @param refreshTimeout - The timestamp after which the refresh attempt will fail, throwing an exception.
+ * @returns - A promise that, if it resolves, will resolve with the T type.
+ */
+async function retryUntilTimeout<T>(
+  asyncFn: () => Promise<T | null>,
+  retryIntervalInMs: number,
+  retryTimeout: number,
+  errorMessage: string
+): Promise<T> {
+  let result = await tryUntilTimeout<T>(asyncFn, retryTimeout, errorMessage);
 
-  while (token === null) {
+  while (result === null) {
```

Comment:
The issue I have with this is that it's still particular to the contract of `getToken` in core-auth. You wouldn't be able to use this abstraction with something that doesn't return `null` as a failure mode. The way this reads to me is "retry every X milliseconds as long as this function is returning null," and then in the inner function we catch errors and return them as null to retry, but that might not always be safe to do in the general case.

If we want to do this refactor with the intention of making this cycler abstraction easier to use with other types of functions, then I think we have to re-think the failure modes to come up with something truly generic. For example, what if the `retryUntilTimeout` function accepts as a parameter a predicate that can be used to determine whether an Error can be retried or should be treated as fatal?

Reply:
I agree that the failure modes are an issue on themselves and that null as a failure mode is not generic, but null as a failure mode is easy for two reasons:

- Has way less overhead.
- It's as clear as possible on the type that uses null as the failure mode. It has been clear enough for us to use it in the TokenCredential, for example 🤔

As an alternative to null as  a failure mode, what would be better while still not make it much more complicated?